feat: Added CalDAV URL to the Accountpage

tammi-23 opened the pull request:
Added a Calendar Panel to the User Account Page.
This includes: CalDAV URL, username and an information about the Calendar-Implementaion.

This implements: https://github.com/opencloud-eu/web/issues/600

tbsbdr commented:
forwarding screenies from @tammi-23 
![image](https://github.com/user-attachments/assets/34997407-206e-47b8-b2b1-f838d8f9fda4)

![image](https://github.com/user-attachments/assets/abc3213e-11e8-4620-92e0-ca44e4a578a9)

AlexAndBear commented on packages/web-runtime/src/components/Account/CalDavUrl.vue:
```diff
@@ -0,0 +1,131 @@
+        method: 'OPTIONS'
+      })
+
+      if (response.request.responseURL.includes('/caldav/')) {
```
Why not severurl/caldav/ seems more like a safe condition to

AlexAndBear commented on packages/web-runtime/src/extensions.ts:
```diff
@@ -2,6 +2,7 @@ import { computed, markRaw } from 'vue'
 import { preferencesPanelExtensionPoint, progressBarExtensionPoint } from './extensionPoints'
 import AppTokens from './components/Account/AppTokens.vue'
 import { CustomComponentExtension, LoadingIndicator } from '@opencloud-eu/web-pkg'
+import CalDavUrl from './components/Account/CalDavUrl.vue'
```
Not 100 percent sure about importing this, I’d rather have an extension point not be dependent on a component, maybe you find a better place for the constant.

AlexAndBear commented on packages/web-runtime/src/components/Account/CalDavUrl.vue:
```diff
@@ -0,0 +1,131 @@
+        calDavAvailable.value = true
+      }
+    } catch (error) {
+      console.error('CalDAV check failed:', error)
```
As noted before, I think a console.info is the better option.

tammi-23 commented:
With the latest push, I updated the Radicale/CalDAV check to verify whether the redirect URL includes /caldav/.
→ https://github.com/opencloud-eu/web/issues/600

I also extended OcTag.vue by adding a color prop with a default value and implemented our website's highlight color (#ca8df5) as a custom option.

Additionally, I added a new configuration to our Docker development stack to enable Radicale.

Still missing: Unit Tests

JammingBen commented on docker-compose.yml:
```diff
@@ -101,9 +101,15 @@ services:
-      traefik.http.routers.opencloud.middlewares: cors
+      traefik.http.routers.opencloud.middlewares: cors,redirect-caldav
+
+      traefik.http.middlewares.redirect-caldav.redirectregex.regex: '^/.well-known/caldav$'
+      traefik.http.middlewares.redirect-caldav.redirectregex.replacement: '/caldav/'
+      traefik.http.middlewares.redirect-caldav.redirectregex.permanent: true
```
Since we're not accessing radicale via traefik, is this stuff really needed? Same with all the other traefik related lines below.

JammingBen commented on packages/design-system/src/components/OcTag/OcTag.vue:
```diff
@@ -30,6 +30,10 @@ export interface Props {
    * @default false
    */
   rounded?: boolean
+  /**
+   * @docs The color of the tag.
+   */
+  color?: 'default' | 'warning' | 'custom-purple'
```
I'd stick with the color role schema as we know it from material: `primary`, `secondary` (default) and `tertiary`.

JammingBen commented on packages/design-system/src/components/OcTag/OcTag.vue:
```diff
@@ -113,4 +126,9 @@ function $_ocTag_click(event: MouseEvent) {
 }
+
+.oc-tag-color-custom-purple {
+  background-color: #ca8df5;
+  color: white;
```
`background-color` would then e.g. be `var(--oc-role-primary)` and `color` would be `var(--oc-role-on-primary)`.

JammingBen commented on packages/web-runtime/src/components/Account/CalDavUrl.vue:
```diff
@@ -0,0 +1,132 @@
+const copyIcon = 'file-copy'
+const checked = ref(false)
+
+const calDavUrl = computed(() => {
+  return `${configStore.serverUrl}`
+})
```
This doesn't need to be a `computed`. Computed properties are only necessary when values as expected to change, which is not the case here. You could just use `configStore.serverUrl` directly where needed.

JammingBen commented on packages/web-runtime/src/components/Account/CalDavUrl.vue:
```diff
@@ -0,0 +1,132 @@
+
+onMounted(async () => {
+  try {
+    const wellKnownUrl = `${configStore.serverUrl}.well-known/caldav`
```
No need to have the server url here, the client service already does this for you.

```suggestion
    const wellKnownUrl = '.well-known/caldav'
```

JammingBen commented on packages/web-runtime/src/components/Account/CalDavUrl.vue:
```diff
@@ -0,0 +1,132 @@
+        method: 'OPTIONS'
+      })
+      console.info('CalDAV check response:', response.request.responseURL)
+      if (response.request.responseURL.includes(`${configStore.serverUrl}caldav/`)) {
```
The `web-client` package provides a helper for building urls like this. It's a bit less error prone since you don't need to manually care about the slashes.

```suggestion
      if (response.request.responseURL.includes(urlJoin(configStore.serverUrl, 'caldav')) {
```

JammingBen commented on packages/web-runtime/tests/unit/components/Account/CalDavUrl.spec.ts:
```diff
@@ -0,0 +1,118 @@
+import { flushPromises } from '@vue/test-utils'
+import { nextTick } from 'vue'
+import CalDavUrl from '../../../../src/components/Account/CalDavUrl.vue'
+import { vi } from 'vitest'
```
This is available globally, no need to import it.

```suggestion
```

JammingBen commented on packages/web-runtime/tests/unit/components/Account/CalDavUrl.spec.ts:
```diff
@@ -0,0 +1,118 @@
+
+describe('CalDavUrl component', () => {
+  it('renders CalDAV information when available', async () => {
+    const wrapper = mount(CalDavUrl, {
+      global: {
+        plugins: defaultPlugins(),
+        mocks: {
+          ...defaultComponentMocks,
+          $gettext: (msg: string) => msg,
+          authStore: {
+            userContextReady: true
+          }
+        }
+      },
+      props: {
+        checked: true,
+        calDavAvailable: true
+      }
+    })
```
I'd suggest having one method per test file that returns the wrapper. See e.g. how it's done the unit tests for the app tokens: https://github.com/opencloud-eu/web/blob/main/packages/web-runtime/tests/unit/components/Account/AppTokens.spec.ts

JammingBen commented on packages/web-runtime/tests/unit/components/Account/CalDavUrl.spec.ts:
```diff
@@ -0,0 +1,118 @@
+        })
+      }
+    }),
+    useUserStore: () => ({
+      user: ref({
+        onPremisesSamAccountName: 'test-user'
+      })
+    }),
+    useConfigStore: () => ({
+      serverUrl: 'https://example.com/'
+    })
```
Are these mocks really needed? If so, please mock them via the plugins like e.g. here: https://github.com/opencloud-eu/web/blob/main/packages/web-runtime/tests/unit/components/Account/AppTokens.spec.ts#L96

JammingBen reviewed (commented):
A few annotations from my side, but overall it looks quite good already!

JammingBen commented on docker-compose.yml:
```diff
@@ -102,8 +102,14 @@ services:
       traefik.http.routers.opencloud.entrypoints: opencloud
       traefik.http.services.opencloud.loadbalancer.server.port: 9200
       traefik.http.routers.opencloud.middlewares: cors
+      #traefik.http.routers.opencloud.middlewares: cors,redirect-caldav
+      #traefik.http.middlewares.redirect-caldav.redirectregex.regex: '^/.well-known/caldav$'
+      #traefik.http.middlewares.redirect-caldav.redirectregex.replacement: '/caldav/'
+      #traefik.http.middlewares.redirect-caldav.redirectregex.permanent: true
+
```
```suggestion
```

JammingBen commented on docker-compose.yml:
```diff
@@ -243,6 +249,9 @@ services:
       traefik.http.middlewares.cors.headers.accesscontrolexposeheaders: '*'
       traefik.http.middlewares.cors.headers.accesscontrolmaxage: 100
       traefik.http.middlewares.cors.headers.addvaryheader: true
+      traefik.http.middlewares.radicale-redirect.redirectregex.regex: '^http://(.*):5232(.*)'
+      traefik.http.middlewares.radicale-redirect.redirectregex.replacement: 'http://$1:9201$2'
+      traefik.http.middlewares.radicale-redirect.redirectregex.permanent: true
```
I think these can be removed as well?

```suggestion
```

JammingBen commented on docker-compose.yml:
```diff
@@ -273,6 +282,25 @@ services:
+      - '5232:5232'
+    volumes:
+      - ./dev/docker/radicale/data:/data
+    #labels:
+    #traefik.enable: true
+    #traefik.http.routers.radicale.tls: true
+    #traefik.http.routers.radicale.rule: PathPrefix(`/`)
+    #traefik.http.routers.radicale.entrypoints: websecure
+    #traefik.http.services.radicale.loadbalancer.server.port: 5232
```
```suggestion
```

AlexAndBear commented on packages/web-runtime/src/components/Account/CalDavUrl.vue:
```diff
@@ -0,0 +1,128 @@
+const calDavAvailable = ref(false)
+const copiedIcon = 'check'
+const copyIcon = 'file-copy'
+const checked = ref(false)
```
Maybe we can give the variable a more descriptive naming

AlexAndBear commented on packages/web-runtime/src/components/Account/CalDavUrl.vue:
```diff
@@ -0,0 +1,128 @@
+        calDavAvailable.value = true
+      }
+    } catch (error) {
+      console.error('CalDAV check failed:', error)
```
```suggestion
      console.info('CalDAV check failed:', error)
```

Since CalDav is not required for a working OpenCloud Instance, I wouldn't announce an error

AlexAndBear reviewed (approved):
🎉

AlexAndBear commented on packages/web-runtime/src/components/Account/CalDavUrl.vue:
```diff
@@ -0,0 +1,128 @@
+      <oc-table-tr class="account-page-info-caldav-url">
+        <oc-table-td>{{ $gettext('CalDAV URL') }}</oc-table-td>
+        <oc-table-td>
+          <span class="oc-text-truncate">{{ configStore.serverUrl }}</span>
```
Did this change ? I thought we need a full url with /caldav slug, now its only the server url ?

kulmann replied:
Only the server URL is correct, because calendar apps will use the .well-known endpoints automagically.
